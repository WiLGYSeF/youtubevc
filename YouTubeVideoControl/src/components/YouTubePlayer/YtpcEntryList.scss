.entry-list {
  background-color: #444;
  border: 1px solid #aaa;
  border-radius: 6px;
  color: #ddd;
  height: 200px;
  overflow: auto;
}
