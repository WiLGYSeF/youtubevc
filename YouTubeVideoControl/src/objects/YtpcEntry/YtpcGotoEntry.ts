import { YouTubePlayer } from 'youtube-player/dist/types';

import YouTubePlayerControllerEntry, { ControlType } from './YouTubePlayerControllerEntry';
import secondsToTimestamp from '../../utils/secondsToTimestamp';

export interface YtpcGotoState {
  goto: number;
}

class YtpcGotoEntry extends YouTubePlayerControllerEntry {
  public static ACTION_STR: string = 'go to';

  public gotoTime: number;

  constructor(atTime: number, gotoTime: number) {
    super(ControlType.Goto, atTime);

    this.gotoTime = gotoTime;
  }

  public performAction(ytPlayer: YouTubePlayer, currentTime: number): void {
    ytPlayer.seekTo(this.gotoTime, true);
  }

  public getActionStr(): string {
    return YtpcGotoEntry.ACTION_STR;
  }

  public getControlStr(): string {
    return secondsToTimestamp(this.gotoTime);
  }
}

export default YtpcGotoEntry;
