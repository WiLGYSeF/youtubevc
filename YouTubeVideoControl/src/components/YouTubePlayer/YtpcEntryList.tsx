import React from 'react';

import YouTubePlayerControllerEntry from 'objects/YtpcEntry/YouTubePlayerControllerEntry';
import YtpcEntry from './YtpcEntry';
import YtpcEntryBar from './YtpcEntryBar';

import styles from './YtpcEntryList.module.scss';

interface YtpcEntryListProps {
  entries: YouTubePlayerControllerEntry[];
  barIndex?: number;
  deleteEntry(entry: YouTubePlayerControllerEntry): void;
  editEntry(entry: YouTubePlayerControllerEntry): void;
}

function YtpcEntryList(props: YtpcEntryListProps) {
  const elements = props.entries.map((e, i) => (
    <YtpcEntry
      key={e.getKey()}
      entry={e}
      deleteEntry={props.deleteEntry}
      editEntry={props.editEntry}
    />
  ));

  if (props.barIndex !== undefined) {
    elements.splice(
      props.barIndex,
      0, (
        <YtpcEntryBar key="entrybar" />
      ),
    );
  }

  return (
    <div className={styles['entry-list']}>
      {elements}
    </div>
  );
}

export function getEntries(container: Element): Element[] {
  return Array.from(container.querySelectorAll('.entry'));
}

export default YtpcEntryList;
