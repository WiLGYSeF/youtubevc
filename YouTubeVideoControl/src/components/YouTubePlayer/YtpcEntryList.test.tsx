import React from 'react';
import renderer from 'react-test-renderer';
import { render } from '@testing-library/react';

import YtpcGotoEntry from 'objects/YtpcEntry/YtpcGotoEntry';
import YtpcEntryList from './YtpcEntryList';

function getFiberNode(node: Node): any {
  const fiberKey = Object.keys(node).find((k) => k.startsWith('__reactFiber$'));
  if (!fiberKey) {
    return null;
  }
  return (node as any)[fiberKey];
}

function getFiberNodeName(node: Node): string | null {
  const fiber = getFiberNode(node);
  if (!fiber) {
    return null;
  }

  const component = fiber.return.elementType as Function;
  return component.name;
}

describe('YtpcEntryList', () => {
  it('renders entries', () => {
    const entries = [
      new YtpcGotoEntry(0, 0),
      new YtpcGotoEntry(5, 0),
      new YtpcGotoEntry(8, 0),
    ];

    const component = renderer.create(<YtpcEntryList
      entries={entries}
      deleteEntry={() => { }}
      editEntry={() => {}}
    />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it.each([
    [0],
    [2],
    [3],
  ])(
    'renders entries with bar at index %d',
    (barIndex: number) => {
      const entries = [
        new YtpcGotoEntry(0, 0),
        new YtpcGotoEntry(5, 0),
        new YtpcGotoEntry(8, 0),
      ];

      const { container } = render(<YtpcEntryList
        entries={entries}
        barIndex={barIndex}
        deleteEntry={() => { }}
        editEntry={() => { }}
      />);

      const children = container.getElementsByClassName('entry-list')[0].childNodes;
      let idx = 0;

      for (; idx < barIndex; idx += 1) {
        expect(getFiberNodeName(children[idx])).toEqual('YtpcEntry');
      }

      expect(getFiberNodeName(children[idx])).toEqual('YtpcEntryBar');
      idx += 1;

      for (; idx < children.length; idx += 1) {
        expect(getFiberNodeName(children[idx])).toEqual('YtpcEntry');
      }
    },
  );
});
