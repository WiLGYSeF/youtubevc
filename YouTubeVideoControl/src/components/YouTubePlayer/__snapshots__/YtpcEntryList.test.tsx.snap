// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`YtpcEntryList renders entries 1`] = `
<div
  className="entry-list"
>
  <div
    className="entry"
  >
    <span>
      At 00:00, go to 00:00
    </span>
    <div
      className="buttons"
    >
      <div
        className="edit"
        onClick={[Function]}
      >
        e
      </div>
      <div
        className="delete"
        onClick={[Function]}
      >
        x
      </div>
    </div>
  </div>
  <div
    className="entry"
  >
    <span>
      At 00:05, go to 00:00
    </span>
    <div
      className="buttons"
    >
      <div
        className="edit"
        onClick={[Function]}
      >
        e
      </div>
      <div
        className="delete"
        onClick={[Function]}
      >
        x
      </div>
    </div>
  </div>
  <div
    className="entry"
  >
    <span>
      At 00:08, go to 00:00
    </span>
    <div
      className="buttons"
    >
      <div
        className="edit"
        onClick={[Function]}
      >
        e
      </div>
      <div
        className="delete"
        onClick={[Function]}
      >
        x
      </div>
    </div>
  </div>
</div>
`;
